### ISSUE TITLE
Cannot decrypt files with pyarrow

### ISSUE THREAD

**Issue by Audrius-GR** [G-Research/ParquetSharp#324]
I am trying to decrypt parquet files that have been produced by ParquetSharp in pyarrow.

How I'm encrypting files:
```
            var columnEncryptionProperties = columns.Select(columnName =>
            {
                var (keyId, aesKeyBytes) = GenerateAesKey(...);
                return new ColumnEncryptionPropertiesBuilder(columnName)
                    .Key(aesKeyBytes)
                    .KeyId(keyId)
                    .Build();
            }).ToArray();
                
            var (footerKeyId, footerKey) = GenerateAesKey(...);

            var fileEncryptionProperties = new FileEncryptionPropertiesBuilder(footerKey)
                .FooterKeyMetadata(footerKeyId)
                .SetPlaintextFooter()
                .EncryptedColumns(columnEncryptionProperties)
                .Build();
            
            var writerProperties = new WriterPropertiesBuilder().Encryption(fileEncryptionProperties)
                .Compression(Compression.Zstd).Build();
            return new ParquetFileWriter(file, schema, writerProperties);
```            

How I'm trying to decrypt files:

Effectively using: 
https://arrow.apache.org/docs/python/parquet.html#parquet-modular-encryption-columnar-encryption
where my KMS client knows how to unwrap keys 

However that fails with:

`OSError: Failed to parse key metadata <base 64 encoded key identity>`

Note that while trying to decrypt, I doesn't get to the point where `unwrap_key` is called in the KMS (or anything on the KMS client at all) , so it's failing to parse the footer.

Looking at arrow code for this, it seems this happens here:

https://github.dev/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/parquet/encryption/key_metadata.cc#L41

Which suggests that the footer metadata should contain not a key identity, but a json which then contains the key identity.

If I open up a parquet file encrypted using pyarrow method in notepad etc, I can indeed see a blob like this in the footer of the file and footer of every column:

```
        {
            "keyMaterialType":"PKMT1",
            "internalStorage":true,
            "isFooterKey":false,
            "masterKeyID":"<key name>",
            "wrappedDEK":"<base 64 encoded key identity>",
            "doubleWrapping":false
        }
```

where as footers produced by ParquestSharp are just:
```
<base 64 encoded key identity>
```

I am not sure how we manage to produce these invalid footers by just using the C++ bindings.

Anyways, as the next step I tried to replace in my C# code that did:
```
.FooterKeyMetadata(footerKeyId)
```

with

```
    public class ParquetKeyMetadata
    {
        [JsonPropertyName("keyMaterialType")] public string KeyMaterialType { get; set; } = "PKMT1";
        [JsonPropertyName("internalStorage")] public bool InternalStorage { get; set; } = true;
        [JsonPropertyName("isFooterKey")] public bool IsFooterKey { get; set; } = false;
        [JsonPropertyName("masterKeyID")] public string MasterKeyID { get; set; }
        [JsonPropertyName("wrappedDEK")] public string WrappedDEK { get; set; }

        [JsonPropertyName("doubleWrapping")] public bool DoubleWrapping { get; set; } = false;

        public static string GenerateSerialized(string masterKeyId, string wrappedDek)
        {
            return JsonSerializer.Serialize(new ParquetKeyMetadata
            {
                MasterKeyID = masterKeyId,
                WrappedDEK = wrappedDek
            });
        }
    }

    ....


    .FooterKeyMetadata(ParquetKeyMetadata.GenerateSerialized("master key", footerKeyId))
```

etc, which matches what arrow expects.

This does produce sort of valid looking parquet files that have json in the footers.

When these files are opened in pyarrow, I do get calls to pythons KMS unwrap (progress), for every key (file footer key and column keys), which suggests arrow does parse the json blob correctly and extracts the key, however it seems the decryption of the actual data fails.

Sometimes it crashes the python kernel, sometimes it fails on the OpenSSL finalizer (OSError: Failed decryption finalization), sometimes it fails at zstd decompression. 

Ironically, these different outcomes all happen for exactly the same file that is using the same keys, so something unexpected/non-deterministic is going on in arrow.

Maybe I started going in the wrong direction (by trying to write out the json blob myself), but I feel at this point I can't make any more progress myself as my ability to navigate arrow codebase and the bindings code is quite limited.

Note that if I try to decrypt a file produced by pyarrow via ParquestSharp, the DecryptionKeyRetriever.FetchKey gets the whole json blob, rather than just key identity, but if you handle parsing of that, it seems to work.

**Comment by adamreeve** [comment]
Hi @Audrius-GR, the main source of the issue here is that ParquetSharp exposes the lower-level encryption API available in the Arrow C++library, but PyArrow only exposes a higher level KMS based API. This higher level API is also available in the Arrow C++ library so could potentially be wrapped by ParquetSharp to make interoperability with PyArrow more straightforward. There was also some discussion about making the lower level API available from Python in https://github.com/apache/arrow/pull/9631, but the conclusion there was that the higher level API is flexible enough that it should be suitable for all use-cases.

The [Parquet format documentation on key metadata](https://github.com/apache/parquet-format/blob/master/Encryption.md#43-key-metadata) doesn't specify that the metadata should be JSON, but considers it to be an arbitrary byte array that developers can use to integrate with a KMS. From a quick look it appears that the JSON metadata used by PyArrow's high level API is compatible with the Java parquet-mr implementation though.

It should be possible to use the lower level  API to produce files compatible with PyArrow. I think you're on the right track by writing JSON data in the footer metadata but you're missing also writing JSON for the column encryption metadata. I was able to read an encrypted file written by ParquetSharp from PyArrow with the following code, which just base64 encodes the keys rather than using a KMS (so this is not secure at all and not suitable for use in production code):

Writing:
```c#
using ParquetSharp;

var key0 = new byte[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
var columns = new Column[]
{
    new Column<int>("x"),
};
var columnEncryptionProperties = columns.Select(column =>
{
    using var columnEncryptionBuilder = new ColumnEncryptionPropertiesBuilder(column.Name);
    return columnEncryptionBuilder 
        .Key(key0)
        .KeyMetadata(
            "{\"keyMaterialType\":\"PKMT1\"," +
            "\"internalStorage\":true," +
            "\"isFooterKey\":false," +
            "\"masterKeyID\":\"key0\"," +
            "\"wrappedDEK\":\"" + System.Convert.ToBase64String(key0) + "\"," +
            "\"doubleWrapping\":false}")
        .Build();
}).ToArray();

using var fileEncryptionBuilder = new FileEncryptionPropertiesBuilder(key0);
using var fileEncryptionProperties = fileEncryptionBuilder
    .FooterKeyMetadata(
        "{\"keyMaterialType\":\"PKMT1\"," +
        "\"internalStorage\":true," +
        "\"isFooterKey\":true," +
        "\"masterKeyID\":\"key0\"," +
        "\"wrappedDEK\":\"" + System.Convert.ToBase64String(key0) + "\"," +
        "\"doubleWrapping\":false," +
        "\"kmsInstanceID\":\"DEFAULT\"," +
        "\"kmsInstanceURL\":\"DEFAULT\"" +
        "}")
    .SetPlaintextFooter()
    .EncryptedColumns(columnEncryptionProperties)
    .Build();

using var writerPropertiesBuilder = new WriterPropertiesBuilder();
using var writerProperties = writerPropertiesBuilder
    .Encryption(fileEncryptionProperties)
    .Compression(Compression.Snappy)
    .Build();

using var writer = new ParquetFileWriter(
    System.Environment.GetCommandLineArgs()[1],
    columns,
    writerProperties);

using var rowGroup = writer.AppendRowGroup();
using var colWriter = rowGroup.NextColumn().LogicalWriter<int>();
colWriter.WriteBatch(new []{0, 1, 2, 3, 4, 5, 6, 7, 8, 9});
writer.Close();
return 0;
```

Reading:
```python
import sys
import base64
import pyarrow.parquet as pq
import pyarrow.parquet.encryption as pe


class MyKmsClient(pe.KmsClient):
   def __init__(self, kms_connection_configuration):
      pe.KmsClient.__init__(self)

   def wrap_key(self, key_bytes, master_key_identifier):
      return base64.b64encode(key_bytes)

   def unwrap_key(self, wrapped_key, master_key_identifier):
      return base64.b64decode(wrapped_key)


def kms_client_factory(kms_connection_configuration):
   return MyKmsClient(kms_connection_configuration)


crypto_factory = pe.CryptoFactory(kms_client_factory)
config = pe.KmsConnectionConfig()
decryption_properties = crypto_factory.file_decryption_properties(config)

file = pq.ParquetFile(sys.argv[1], decryption_properties=decryption_properties)
table = file.read()
print(table)
```

**Comment by Audrius-GR** [comment]
So the example that I posted that I claimed didn't work, actually works fine today.
The only thing that has changed over the weekend is my machine having been restarted.

I suspect I saw crashes because I had reinstalled the version of pyarrow to match that of this package, yet the dll was already loaded somewhere, so I ended up with pyx files for 10.0.1, and the native dll for some other version resulting in crashes.

Anyways, closing as the workaround of serializing the json ourselves works.
